exercises/circular-buffer/canonical-data.json

```diff
@@ -0,0 +1,296 @@
+{
+  "#": [
+    "In general, these circular buffers are expected to be stateful,",
+    "and each language will operate on them differently.",
+    "Tests tend to perform a series of operations, some of which expect a certain result.",
+    "As such, this common test suite can only say in abstract terms what should be done.",
+    "",
+    "Tests will contain a number of operations. The operation will be specified in the `operation` key.",
+    "Based on the operation, other keys may be present.",
+    "read: Reading from the buffer should produce the item located at `expected`.",
+    "write: Writing the item located at `item` should succeed.",
+    "readShouldFail: Reading from the buffer should fail.",
+    "writeShouldFail: Writing the item located at `item` should fail.",
+    "overwrite: Write the item located at `item` into the buffer, removing the oldest item if necessary.",
+    "clear: Clear the buffer.",
+    "",
+    "Finally, note that all values are integers.",
+    "If your language contains generics, you may consider allowing buffers to contain other types.",
+    "Tests for that are not included here.",
+    ""
+  ],
+  "cases": [
+    {
+      "description": "reading empty buffer should fail",
+      "capacity": 1,
+      "operations": [
+        {
+          "operation": "readShouldFail"
```

Comment:
Command Query Responsibility Segregation.

write is a command.
overwrite is a command.
clear is a command.
read is a query.

detecting failure is a query.

It seems we are missing the `failure_detected`\* query that indicates this.

```
  "cases": [
    {
      "description": "reading empty buffer should fail",
      "capacity": 1,
      "operations": [
        {
          "operation": "read",
          "expected": null
        },
        {
          "operation": "failure_detected",
          "expected": true
        }
      ]
    },
```

Individual tracks are still free to implement the `failure_detected` check however they like, either with a method or a try/catch block or whichever other system their language encourages.

\* There's probably a better name for this.